# Task
Consigne 2 : Modélisation en NLP

Here is all the data you need:
"submission.csv"

## Data loading

### Subtask:
Load the data from "submission.csv" into a pandas DataFrame.


**Reasoning**:
Load the data from "submission.csv" into a pandas DataFrame and display the first few rows.



In [1]:
import pandas as pd

try:
    df = pd.read_csv('submission.csv')
    display(df.head())
except FileNotFoundError:
    print("Error: 'submission.csv' not found. Please make sure the file exists in the current directory.")
    df = None  # Set df to None to indicate failure
except Exception as e:
    print(f"An unexpected error occurred: {e}")
    df = None

,textID,selected_text
0,f87dea47db,Last session of the day http://twitpic.com/67ezh
1,96d74cb729,exciting
2,eee518ae67,Recession
3,01082688c6,happy bday!
4,33987a8ee5,I like it!!


## Data exploration

### Subtask:
Explore the loaded data to understand its structure, identify the relevant text column(s) for NLP, and check for missing values or inconsistencies. Determine the target variable if applicable.


**Reasoning**:
Examine the DataFrame's shape and info, identify text columns, check for missing values, explore the target variable (if applicable), inspect unique values in the text columns, and summarize findings.



In [2]:
# Examine the DataFrame's shape and info
print("DataFrame Shape:", df.shape)
print("\nDataFrame Info:")
print(df.info())

# Identify text columns and check for missing values
text_columns = ['selected_text']  # Assuming 'selected_text' is the relevant text column

print("\nMissing Values:")
print(df[text_columns].isnull().sum())

# Explore the target variable (Not applicable in this case as no target variable is explicitly provided)

# Inspect unique values
for col in text_columns:
    print(f"\nUnique values in '{col}': {df[col].nunique()}")
    print(f"Examples of text in '{col}': {df[col].unique()[:10]}")

# Summarize findings
print("\nSummary:")
print("The DataFrame contains", df.shape[0], "rows and", df.shape[1], "columns.")
print("The 'selected_text' column appears to be the relevant text column for NLP.")
print("There are no missing values in the text column. ")
# Note: No target variable was found in the provided data.
# This needs to be clarified in the problem description for accurate NLP model training.

DataFrame Shape: (3534, 2)

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3534 entries, 0 to 3533
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   textID         3534 non-null   object
 1   selected_text  3534 non-null   object
dtypes: object(2)
memory usage: 55.3+ KB
None

Missing Values:
selected_text    0
dtype: int64

Unique values in 'selected_text': 2977
Examples of text in 'selected_text': ['Last session of the day  http://twitpic.com/67ezh' 'exciting'
 'Recession' ' happy bday!' 'I like it!!' 'visitors!' 'HATES' 'blocked'
 ' and within a short time of the last clue all of them'
 ' What did you get?  My day is alright.. haven`t done anything yet. leaving soon to my stepsister though!']

Summary:
The DataFrame contains 3534 rows and 2 columns.
The 'selected_text' column appears to be the relevant text column for NLP.
There are no missing values in the text column. 


## Data preparation

### Subtask:
Clean the text data in the 'selected_text' column of the DataFrame `df`.


**Reasoning**:
Clean the 'selected_text' column by lowercasing, removing punctuation, handling special characters, removing extra whitespace, and optionally stemming or lemmatizing.  Create a new 'cleaned_text' column to store the results. I will use a simple approach for special character handling (replacing with spaces) and will not perform stemming/lemmatization in this initial attempt to avoid unnecessary complexity at this stage.



In [3]:
import re

def clean_text(text):
    """Cleans a given text string."""
    if isinstance(text, str):  # Check if the input is a string
        text = text.lower()
        text = re.sub(r'[^\w\s]', ' ', text)  # Remove punctuation
        text = re.sub(r'\s+', ' ', text)  # Remove extra whitespace
        text = text.strip()
        return text
    else:
        return ''  # Handle non-string inputs

# Apply the cleaning function to the 'selected_text' column and create 'cleaned_text'
df['cleaned_text'] = df['selected_text'].apply(clean_text)
display(df.head())

,textID,selected_text,cleaned_text
0,f87dea47db,Last session of the day http://twitpic.com/67ezh,last session of the day http twitpic com 67ezh
1,96d74cb729,exciting,exciting
2,eee518ae67,Recession,recession
3,01082688c6,happy bday!,happy bday
4,33987a8ee5,I like it!!,i like it


## Feature engineering

### Subtask:
Extract TF-IDF features from the cleaned text data.


**Reasoning**:
Extract TF-IDF features from the 'cleaned_text' column using TfidfVectorizer.



In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Initialize TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')

# Fit and transform the 'cleaned_text' column
tfidf_matrix = tfidf_vectorizer.fit_transform(df['cleaned_text'])

# Print the shape of the tfidf matrix
print(tfidf_matrix.shape)

(3534, 5000)


## Data splitting

### Subtask:
Split the TF-IDF matrix into training and testing sets.


**Reasoning**:
Split the TF-IDF matrix into training and testing sets using train_test_split.



In [5]:
from sklearn.model_selection import train_test_split

# Assuming tfidf_matrix is already created from the previous step
X_train, X_test, _, _ = train_test_split(tfidf_matrix, [0] * tfidf_matrix.shape[0], test_size=0.2, random_state=42)
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")

X_train shape: (2827, 5000)
X_test shape: (707, 5000)


## Model training

### Subtask:
Train a Logistic Regression model on the training data (X_train) and evaluate its performance on the test data (X_test).  Since there's no target variable provided in the original dataset, this will be a demonstration of the model training process, and no actual prediction or evaluation can be performed.


**Reasoning**:
Train a Logistic Regression model on the training data and evaluate its performance on the test data. Create dummy target variables since the actual target is missing.



In [7]:
from sklearn.linear_model import LogisticRegression
import numpy as np

# Initialize the Logistic Regression model
model = LogisticRegression(random_state=42)

# Create a dummy target variable with two classes
y_train = np.random.randint(0, 2, size=X_train.shape[0])

# Train the model
model.fit(X_train, y_train)

# Create a dummy target variable for the test set
y_test = np.random.randint(0, 2, size=X_test.shape[0])

# Predict on the test data (not meaningful without a real target variable)
y_pred = model.predict(X_test)

# Print the predicted values
print(y_pred)

[0 1 0 1 1 1 1 1 1 1 1 1 0 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 0 1 0 0 1 0 0
 0 0 1 1 1 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 0 1 1 0 0 1 0 1 1 0 0 1 1 1
 0 0 1 1 0 0 0 0 1 0 1 1 0 0 0 0 1 1 0 1 1 0 1 1 1 0 0 0 1 1 1 0 0 0 0 1 1
 0 1 1 0 0 0 0 1 1 1 0 0 0 0 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 0 0 0 0 0 1 0 1
 0 1 0 1 0 1 0 1 0 1 0 0 1 1 0 0 0 0 1 1 0 1 1 1 1 0 0 0 0 0 1 0 1 0 1 1 0
 0 0 0 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1 1 1 0 0 1 0
 0 1 0 1 0 1 0 0 0 1 0 0 1 0 1 1 0 0 0 1 1 1 1 1 1 0 1 1 0 1 0 0 0 0 1 1 1
 1 1 1 0 1 0 1 0 0 1 1 0 1 0 0 1 0 0 0 1 1 1 1 0 1 0 1 1 0 1 0 1 1 1 1 1 1
 1 1 1 1 1 1 0 1 0 1 1 1 1 1 0 0 1 1 1 1 0 0 0 0 1 0 1 0 1 1 1 0 1 0 0 1 1
 0 1 1 0 1 0 0 0 1 0 0 1 1 1 1 1 1 1 1 0 0 0 1 1 1 0 0 0 1 0 1 1 1 1 1 0 1
 0 1 1 1 1 1 1 0 1 1 0 1 1 0 0 1 1 1 1 1 1 1 0 0 1 1 0 1 1 1 0 1 1 1 0 1 1
 1 0 1 1 0 1 0 1 0 1 0 0 1 1 0 1 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 0 0 1 1 1 1
 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0 0 1 0 1 0
 1 1 0 1 1 0 1 1 0 0 1 0 

## Model optimization

### Subtask:
Optimize the Logistic Regression model's hyperparameters using GridSearchCV.


**Reasoning**:
Optimize the Logistic Regression model's hyperparameters using GridSearchCV with a defined parameter grid and appropriate scoring metric.



In [8]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Define the parameter grid
param_grid = {'C': [0.1, 1, 10], 'penalty': ['l2'], 'solver': ['liblinear', 'saga']}

# Instantiate the Logistic Regression model
model = LogisticRegression(random_state=42)

# Instantiate GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, scoring='accuracy', cv=5)

# Fit GridSearchCV to the training data
grid_search.fit(X_train, y_train)

# Print the best hyperparameters and score
print(f"Best Hyperparameters: {grid_search.best_params_}")
print(f"Best Score: {grid_search.best_score_}")

Best Hyperparameters: {'C': 0.1, 'penalty': 'l2', 'solver': 'saga'}
Best Score: 0.5037136871071641


## Model evaluation

### Subtask:
Evaluate the performance of the optimized Logistic Regression model on the test data (`X_test`, `y_test`)


**Reasoning**:
Evaluate the performance of the optimized Logistic Regression model on the test data (`X_test`, `y_test`).  Since `y_test` was created as a dummy variable in the previous step, the evaluation will not be meaningful but will follow the instructions.



In [9]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Use the best estimator from the GridSearchCV object
best_model = grid_search.best_estimator_

# Predict on the test data
y_pred = best_model.predict(X_test)

# Calculate the metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print the metrics
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")
print(f"Confusion Matrix:\n{conf_matrix}")

Accuracy: 0.5190947666195191
Precision: 0.5223048327137546
Recall: 0.771978021978022
F1-score: 0.623059866962306
Confusion Matrix:
[[ 86 257]
 [ 83 281]]


## Summary:

### 1. Q&A

No explicit questions were provided in the data analysis task.  However, the analysis implicitly addresses several questions:

* **What is the structure of the data?** The data has 3534 rows and 2 columns ('textID', 'selected_text').  The 'selected_text' column is the relevant text data for NLP.
* **Are there missing values?** No missing values were found in the 'selected_text' column.
* **What is the appropriate NLP task for this data?**  The absence of a target variable makes it unclear.  The data is suitable for unsupervised tasks or requires a target variable for supervised learning.
* **How can we represent the text data for modeling?** TF-IDF features were extracted.
* **What is the performance of the model?**  The model's performance was evaluated using accuracy, precision, recall, F1-score, and a confusion matrix.  However, the results are based on dummy data, making them unreliable.

### 2. Data Analysis Key Findings

* **No Target Variable:** The dataset lacks a target variable, hindering supervised learning tasks and making the model training and evaluation results unreliable.  The analysis highlights this crucial missing piece of information.
* **Text Cleaning:** The 'selected_text' column was cleaned by lowercasing, removing punctuation, handling special characters, and removing extra whitespace. The cleaned text is stored in a new 'cleaned_text' column.
* **TF-IDF Feature Extraction:** TF-IDF features were extracted from the cleaned text data. The resulting TF-IDF matrix has a shape of (3534, 5000), indicating 5000 features were extracted from 3534 text samples.  The vectorizer used `max_features=5000`, `ngram_range=(1, 2)`, and `stop_words='english'`.
* **Model Optimization:** Hyperparameter tuning using GridSearchCV identified the best hyperparameters as `{'C': 0.1, 'penalty': 'l2', 'solver': 'saga'}` achieving a best cross-validation accuracy score of 0.5037 (using dummy target variables).
* **Model Evaluation (with dummy data):** The optimized model (using dummy target variables) achieved an accuracy of 0.519, precision of 0.522, recall of 0.772, and an F1-score of 0.623 on the test set.  The confusion matrix was also reported, but these results are not meaningful without a real target variable.

### 3. Insights or Next Steps

* **Define a clear NLP task and acquire a target variable:**  The most crucial next step is to define the specific NLP task (e.g., sentiment analysis, text classification, etc.) and obtain a corresponding target variable for supervised learning. Without a target variable, meaningful model training and evaluation are impossible.
* **Explore alternative feature engineering techniques:** While TF-IDF is a common approach, consider other feature engineering methods (e.g., word embeddings like Word2Vec or GloVe) or explore advanced NLP models (e.g., transformers) to potentially improve performance once a target variable is available.
